### PR TITLE
[ROCm] Fix hipify skipping sources under Windows subst drives/symlinks in CUDAExtension

### PR DESCRIPTION
**Pull request by pablo-garay**
﻿## Summary

`CUDAExtension` scopes hipify to `os.getcwd()` and filters sources with `includes=[build_dir/*]`, comparing against `os.path.abspath(...)` source paths. When a build passes source paths that resolve through symlinks/junctions or Windows `subst` drive aliases, the two sides end up with different spellings of the same location, the `includes` filter matches nothing, and hipify **silently skips** the GPU sources leaving raw CUDA headers (e.g. `cuda_runtime_api.h`) for `hipcc`, which then fails with `fatal error: 'cuda_runtime_api.h' file not found`.

This reproduces on Windows ROCm CI, where the build runs under a `subst` drive (e.g. `B:`) while a dependency such as torchaudio computes its source directory with `Path(...).resolve()` and lands on the real drive (e.g. `C:`). `build_dir` (`B:\...`) and the source paths (`C:\...`) then disagree.

## Fix

The root cause is in `torch/utils/hipify/hipify_python.py`: `hipify()` appended `extra_files` to `all_files` verbatim, while `preprocessor()` tests `_to_unix_path(filepath)` against `all_files`. On Windows the appended backslash spelling never equals the forward-slash form being tested, so the `extra_files` escape hatch was effectively dead and hipification depended entirely on the `includes=[build_dir/*]` directory walk — which the subst-drive mismatch defeats. `extra_files` entries are now normalized with `_to_unix_path` on append (and added to `all_files_set`), fixing the silent skip at the source in a platform-independent, unit-testable way.

In `CUDAExtension` (and the matching `load()` / `_jit_compile` path), each source is canonicalized only when its plain `os.path.abspath` isn't already under `build_dir`. This fixes the subst-drive case without `realpath`-ing sources that are legitimately symlinked *into* the build dir — doing so would relocate the generated `.hip` outside `build_dir` and break the "paths relative to the setup.py directory" invariant. The build directory itself is still `realpath`'d (a no-op on POSIX).

Finally, when a `.cu`/`.cuh` source comes back unhipified, we now log a warning so this failure mode is self-diagnosing instead of surfacing later as a missing `cuda_runtime_api.h`.

The only Linux-visible behavior change is the guarded source canonicalization; the build-dir `realpath` and the `extra_files` normalization are no-ops on POSIX.

## Test plan

- New unit test `test/test_utils.py::TestHipify::test_hipify_keeps_symlinked_source_under_build_dir` (runs on Linux, no ROCm hardware): a `.cu` symlinked into the build dir is hipified in place, with the generated `.hip` staying under the build dir and not following the symlink into the real source tree.
- Windows ROCm extension build (e.g. torchaudio) hipifies and compiles GPU sources: the hipify pass reports kernel replacements > 0 and produces `.hip` sources instead of leaving `.cu`.
- Trunk CI green on this PR head (all checks passed): [trunk run](https://github.com/pytorch/pytorch/actions/runs/32807838458) · [HUD](https://hud.pytorch.org/pr/194641)

Cherry-pick of ROCm/pytorch#3561 (ROCM-29365), validated end-to-end there with a full Windows ROCm multi-arch wheel build (torch + torchaudio + torchvision).

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Review by jithunnair-amd — APPROVED** [review]
@jeffdaily Please review

**Review by jeffdaily — CHANGES_REQUESTED** [review]
Thanks for chasing this down; the `subst`-drive diagnosis matches what I'd expect from the `includes` scope filter. A few things need to change before this goes in.

**1. `realpath` on the sources changes Linux behavior and relocates generated files.**

On POSIX `os.getcwd()` is already fully resolved, so `os.path.realpath(os.getcwd())` is a no-op on Linux. The only Linux-visible change here is `os.path.realpath(s)` over `sources`, so "No effect on Linux" isn't quite right. When a source inside the build dir is a symlink (build systems that overlay/symlink sources into the build tree do this), `realpath` moves it out of `build_dir`. Repro against current `hipify_python` with `build/foo.cu -> ../real/foo.cu`:

```
MODE abspath   key=/tmp/hipsym/build/foo.cu  -> /tmp/hipsym/build/foo.hip [ok]
MODE realpath  key=/tmp/hipsym/real/foo.cu   -> /tmp/hipsym/real/foo.hip  [ok]
```

The generated `.hip` now lands in the source tree rather than the build dir, and `os.path.relpath(hipified_s_abs, build_dir)` then yields `../real/foo.hip`, which is exactly what the comment on the next line rules out ("must *always* be /-separated paths relative to the setup.py directory"). Nothing cleans those up either, since this call site passes no `clean_ctx`.

Suggest canonicalizing only when the plain `abspath` isn't already under the build dir:

```python
def _canonical_source(s):
    s_abs = os.path.abspath(s)
    return s_abs if s_abs.startswith(build_dir + os.sep) else os.path.realpath(s)
```

That still fixes the reported case (torchaudio passes `C:\...` absolutes while cwd is `B:\...`, so the prefix test fails and `realpath` applies) without moving symlinked sources.

**2. The real root cause is in `hipify_python`, and this only fixes sources that live under `build_dir`.**

`hipify()` appends `extra_files` to `all_files` verbatim, while `preprocessor()` tests `_to_unix_path(filepath)` against `all_files`. Since `_to_unix_path` replaces `os.sep`, on Windows the appended backslash spelling can never match the forward-slash form being tested, so the `extra_files` escape hatch is effectively dead and hipification depends entirely on the directory walk matching `includes=[build_dir/*]`. That's why the mismatch showed up as a silent skip. Normalizing on append (`all_files.append(_to_unix_path(f))`, plus the `all_files_set` membership test) fixes it at the source, is platform-independent, and unlike the change here it's unit-testable without ROCm hardware. As written, any Windows source outside `build_dir` is still silently skipped.

**3. `load()` has the identical pattern and isn't touched.**

`torch/utils/cpp_extension.py` around lines 2380/2390 uses the same `extra_files=[os.path.abspath(s) for s in sources]` / `s_abs = os.path.abspath(source)` pair with `project_directory=build_directory`. Under the same setup (build dir on the `subst` drive, caller-supplied sources resolved to the real drive) `load`/`load_inline` should skip hipify too. Either apply the same treatment or say why it's unaffected.

**4. Please make the skip loud.**

The whole failure mode is "hipify quietly did nothing" and the user only learns about it when `hipcc` reports a missing `cuda_runtime_api.h`. Warning in the loop when a `.cu`/`.cuh` source comes back with `hipified_path is None` (or missing from `hipify_result`) would make this class of bug self-diagnosing.

**5. Test.**

The symlink regression in (1) reproduces on Linux with no ROCm hardware; `torch.utils.hipify.hipify_python` is pure Python and importable standalone, and `test/test_utils.py` already has hipify tests. A test that puts a symlinked `.cu` in a build dir, calls `hipify(..., includes=[build_dir/'*'], extra_files=[...], hipify_extra_files_only=True)`, and asserts the output stays under the build dir would pin both the fix and the regression. If the normalization moves into `hipify_python` per (2), the same test covers the path-spelling case as well. Trunk being green doesn't cover any of this, since no CI job builds under a `subst` drive.

---
*Reviewed with assistance from Claude (Anthropic); findings verified locally and edited by me.*

**Comment by pablo-garay** [comment]
Thanks for the thorough review — fixing it at the root was the right call. Addressed all points in 86420ca:

1. **Guarded source canonicalization (Linux regression).** Sources are now `realpath`'d only when their plain `os.path.abspath` isn't already under `build_dir`, so a source symlinked into the build dir stays put and the generated `.hip` doesn't leave `build_dir`. `build_dir` itself is still `realpath`'d (a no-op on POSIX). I also corrected the inaccurate "No effect on Linux" wording in the description — the only Linux-visible change is this guarded canonicalization.

2. **Root cause in `hipify_python`.** `hipify()` now normalizes `extra_files` with `_to_unix_path` on append (and adds them to `all_files_set`), so the appended spelling matches the `_to_unix_path` membership test in `preprocessor()`. This revives the `extra_files` escape hatch on Windows and is platform-independent, so any source outside `build_dir` is no longer silently skipped.

3. **`load()` / `_jit_compile`.** Applied the same guarded canonicalization there too, since it had the identical `abspath` pattern with `project_directory=build_directory`.

4. **Loud skip.** Both loops now emit a `logger.warning` when a `.cu`/`.cuh` source comes back with no hipified path, so this class of bug is self-diagnosing instead of surfacing later as a missing `cuda_runtime_api.h`.

5. **Test.** Added `test/test_utils.py::TestHipify::test_hipify_keeps_symlinked_source_under_build_dir`: it symlinks a `.cu` into a build dir, runs `hipify(..., includes=[build_dir/'*'], extra_files=[...], hipify_extra_files_only=True)`, and asserts the `.hip` stays under the build dir and isn't written into the real source tree — reproducible on Linux with no ROCm hardware. The backslash-spelling half of (2) is Windows-only (`_to_unix_path` is a no-op on POSIX), so I verified that path manually on Windows: an out-of-scope `extra_files` source is skipped without the fix and hipified with it.

PTAL.

**Review by jeffdaily — COMMENTED** [review]
Re-reviewed `86420ca`. The `hipify_python` normalization is the right root-cause fix. Three things still need to change.

**1. The new test passes on unmodified `main`.**

`test_hipify_keeps_symlinked_source_under_build_dir` calls `hipify()` directly, so it never reaches `CUDAExtension`'s `_canonical_source`, which is where the "don't realpath a source already under `build_dir`" decision lives. And `_to_unix_path` is the identity on POSIX, so the `hipify_python.py` change is a no-op for it too. Running the test body against current `main`:

```
/tmp/tmp04kh2mnc/build/kernel.cu -> /tmp/tmp04kh2mnc/build/kernel.hip [ok]
TEST PASSED on unpatched main
```

Neither behavior change in the PR is pinned. The Linux-visible half is testable if `_canonical_source` moves to module scope (it is currently duplicated verbatim in `CUDAExtension` and `_jit_compile`): assert it returns the `abspath` for a path under `build_dir` and the `realpath` for one outside. The separator half is Windows-only and I am fine leaving it uncovered, but then the test name and docstring shouldn't read as if they cover it.

**2. The root-cause fix leaves two siblings in the same block broken the same way.**

`hipify_python.py:1165` iterates the raw `extra_files` argument; the loop above rebinds a local `f`, so neither the `os.path.join(output_directory, f)` nor the new `_to_unix_path(f)` reaches the iteration. A relative `extra_files` entry is still silently skipped on every platform after this PR:

```
extra_files=["kernel.cu"], hipify_extra_files_only=True
-> {'/tmp/.../build/kernel.cu': (None, '[ignored, not to be hipified]')}
```

Building the normalized list once and iterating that fixes it.

`hipify_python.py:1152-1157` extends `all_files` with `str(path)` from `Path.rglob`, which is backslash-spelled on Windows and so can never match the `_to_unix_path`-normalized `filepath` at line 831. Additionally `_fnmatch(str(path), includes)` on line 1154 compares that backslash path against `includes`, normalized at line 1130, so on Windows `header_include_dirs` contributes nothing at all. `CUDAExtension` passes `header_include_dirs=include_dirs`, so an extension whose `.cuh` sits in an `include_dirs` entry stays unhipified on Windows even with this PR applied. That is the same failure mode the PR is fixing.

**3. The `_jit_compile` guard is inconsistent with itself.**

It tests `os.path.abspath(s).startswith(os.path.realpath(build_directory) + os.sep)` while leaving `project_directory`/`output_directory` as the unresolved `build_directory`; `CUDAExtension` resolves both. `load_inline` writes `main.cpp`/`cuda.cu` into `build_directory` and passes those as `sources`, so with a symlinked `build_directory=` or `TORCH_EXTENSIONS_DIR` (the default from `get_default_build_root()` is already realpath'd) the prefix test fails for files that genuinely are in the build dir and they take the `realpath` branch, which is exactly what the guard exists to prevent. Compare `os.path.realpath(s)` against the resolved base, or resolve `build_directory` once and pass the resolved form through as `CUDAExtension` does.

Also: the CI-green claim in the description points at run `32807838458`, which predates `86420ca`. Current head has 4 red mi300 jobs, all unrelated as far as I can tell (`test_Transformer_multilayer_coder_cuda_tf32` failing consistently, and two inductor shards hitting the 270 minute job timeout), but the link is worth refreshing.

---
*Reviewed with assistance from Claude (Anthropic); findings verified locally and edited by me.*
